src/main/java/com/teamrocket/tms/services/user/UserServiceImpl.java

```diff
@@ -69,10 +70,26 @@ public UserDTO createUser(UserDTO userDTO) {
 
     @Override
     public UserDTO updateUser(Long userId, UserDTO userDTO) {
-        userRepository.findById(userId).orElseThrow(() -> new UserNotFoundException("User with id " + userId + " not found."));
-
-        User user = modelMapper.map(userDTO, User.class);
-        user.setId(userId);
+        User user = userRepository.findById(userId)
+                .orElseThrow(() -> new UserNotFoundException("User with id " + userId + " not found."));
+
+        if (userDTO.getFirstName() != null) {
+            if (userDTO.getFirstName().length() < 3 || userDTO.getFirstName().length() > 30) {
+                throw new UserPropertiesException("must be between 3 and 30 characters");
+            } else {
+                user.setFirstName(userDTO.getFirstName());
+            }
+        }
+        if (userDTO.getLastName() != null) {
+            if (userDTO.getLastName().length() < 3 || userDTO.getLastName().length() > 30) {
+                throw new UserPropertiesException("must be between 3 and 30 characters");
+            } else {
+                user.setLastName(userDTO.getLastName());
+            }
+        }
+        if (userDTO.getRole() != null) user.setRole(userDTO.getRole());
```

Comment:
User should not be able to update / set own role without any validations:
- Team Leader can change to Project Manager